server/client/index.ts

```diff
@@ -0,0 +1,111 @@
+import { get } from "lodash";
+import {
+  IContextProvider,
+  ILegacyScopedClusterClient,
+  OpenSearchDashboardsRequest,
+  RequestHandler,
+  RequestHandlerContext,
+} from "opensearch-dashboards/server";
+import { IRequestHandlerContentWithDataSource, IGetClientProps, DashboardRequestEnhancedWithContext } from "./interface";
+
+export const getClientSupportMDS = (props: IGetClientProps) => {
+  const originalAsScoped = props.client.asScoped;
+  const handler: IContextProvider<RequestHandler<unknown, unknown, unknown>, "core"> = (
+    context: RequestHandlerContext,
+    request: OpenSearchDashboardsRequest
+  ) => {
+    (request as DashboardRequestEnhancedWithContext)[`${props.pluginId}_context`] = context as IRequestHandlerContentWithDataSource;
```

Comment:
I remember you mentioned the concern of putting `_data_source_id_` in http headers and the challenge of moving it to url search parameters.

Since we modified the original `request` here anyway(though it looks a bit hacky as `registerRouteHandlerContext` is not intent for this). If we end up with modifying the request object, perhaps we can also do this to handle data source id in search parameters:

```javascript
    core.http.registerRouteHandlerContext(`${props.pluginId}_MDS_CTX_SUPPORT`, (ctx, req) => {
      req[`${props.pluginId}_context`] = ctx;
      if (req.url.searchParams.has('__data_source_id__')) {
        // add data source id to request object so later access
        req[`${props.pluginId}_data_source_id`] = req.url.searchParams.get('__data_source_id__');
        // then we can remove it from search params so it won't bother router validators.
        req.url.searchParams.delete('__data_source_id__');
      }
      return {};
    });
```